Bug #1674 q sequence revamp

jan-petr opened the pull request:
### Linked issue

Close #1674

jan-petr committed:
#1674 Correct PulseSequenceType import

jan-petr committed:
#1674 Fix PulseSequenceType BIDS2Legacy conversion

jan-petr committed:
#1674 Rename readoutDim to MRAcquisitionType

jan-petr committed:
#1674 Remove use of x.Q.Sequence

jan-petr committed:
#1674 Fix PulseSequenceType loading

HenkMutsaerts commented on Functions/Administration/xASL_adm_DefineASLSequence.m:
```diff
@@ -5,87 +5,86 @@
-        warning(['Unknown ASL sequence: ' xASL_num2str(xQ.Sequence)]);
+if isfield(xQ, 'PulseSequenceType')
+    if isempty(regexpi(xQ.PulseSequenceType, '(epi|grase|spiral)', 'once'))
+        warning(['Unknown ASL sequence: ' xASL_num2str(xQ.PulseSequenceType)]);
```
should be

warning(['Unknown ASL readout PulseSequenceType: ' xASL_num2str(xQ.PulseSequenceType)]);

jan-petr replied:
Done

HenkMutsaerts commented on Functions/Administration/xASL_adm_DefineASLSequence.m:
```diff
@@ -5,87 +5,86 @@
+    warning('xQ.MRAcquisitionType parameter missing');
 end
 
 % Check for illegal sequence definitions
```
This was done for Sequence which is MRAcquisitionType _ PulseSequenceType. So now we only do this for PulseSequenceType, let's also do this for MRAcquisitionType.

jan-petr replied:
MRAcquisitionType_PulseSequenceType is really used in function xASL_adm_DefineASLResolution - I have restored the original functionality and I am checking for both readout and sequence now.

jan-petr replied:
For all files where x.Q.Sequence was used, we are now checking for both PulseSequenceType and MRAcquisitionType. 

You can resolve.

HenkMutsaerts commented on Functions/Administration/xASL_adm_DefineASLSequence.m:
```diff
@@ -5,87 +5,86 @@
 % FORMAT: [xQ] = xASL_adm_DefineASLSequence(xQ, bVerbose)
```
Now we improve the nomenclature, shouldn't we call this function:

xASL_adm_DefineASLReadout

instead?

jan-petr replied:
It is still defining MRAcquisitionType and PulseSequenceType, so Sequence is equally valid as before. Readout isn't incorrect either, but I don't see why a change would be necessary.

HenkMutsaerts replied:
Because ASL sequence = labeling + readout. All this concerns the readout only. It's not a big deal, but if we 're changing the nomenclature anyway to more correct/clear, then readout is more clear than "sequence".

What sequence is it? A PCASL. Is said as more often than "What type of ASL sequence is it?" "2D_EPI"
So not (in)correct, just more clear.

I don't feel strong about this though :)

jan-petr replied:
Updated to readout.

HenkMutsaerts commented on Functions/Administration/xASL_adm_DefineASLSequence.m:
```diff
@@ -5,87 +5,86 @@
 end
 
 
 %% Try to work out which ASL sequence we have
```
then this would become

%% Try to work out which ASL readout we have

(and similar perhaps throughout this function)

jan-petr replied:
Same as above - nothing has changed about the functionality.

HenkMutsaerts replied:
Not what I meant. I meant that we are checking the readout, not the sequence (sequence being labeling + readout)

jan-petr replied:
I updated to "readout".

HenkMutsaerts commented on Functions/Administration/xASL_adm_DefineASLSequence.m:
```diff
@@ -5,87 +5,86 @@
-		xQ.Sequence = '2D_EPI';
+if ~isfield(xQ, 'PulseSequenceType') && isfield(xQ, 'MRAcquisitionType') 
+	if strcmpi(xQ.MRAcquisitionType, '2D')
+		xQ.PulseSequenceType = 'EPI';
```
shouldn't this be lowercase per BIDS?

I know we used uppercase but are increasingly using strcmpi and regexpi so this is a good check and easy BIDS compatibility.

jan-petr replied:
No. EPI is according to BIDS - I've checked that before implementing.

HenkMutsaerts commented on Functions/Administration/xASL_adm_DefineASLSequence.m:
```diff
@@ -5,87 +5,86 @@
+		xQ.PulseSequenceType = 'EPI';
         if bVerbose
-            if ~isempty(regexpi(xQ.Vendor,'Gold Standard Phantoms'))
+            if ~isempty(regexpi(xQ.Vendor, 'Gold Standard Phantoms', 'once'))
```
We could change xQ.Vendor to the BIDS keyname (Manufacturer) but I would not do that now.
We could though change x.Q.Sequence now to x.Q.Readout or something, which is more correct and also nice for ourselves.

jan-petr replied:
Vendor -> Manufacturer. I agree, I made an issue #1683.

No point in changing x.Q.Sequence to x.Q.Readout as this is not a BIDS fields and we want to stay BIDS compatible -> we use MRAcquisitionType and PulseSequenceType which are BIDS compatible.

HenkMutsaerts commented on Functions/Administration/xASL_adm_DefineASLSequence.m:
```diff
@@ -5,87 +5,86 @@
-			xQ.Sequence = '3D_GRASE'; % assume that 3D Philips or Siemens is 3D GRASE
+	elseif isfield(xQ, 'Vendor') && strcmpi(xQ.MRAcquisitionType, '3D')
+		if  ~isempty(regexpi(xQ.Vendor, 'Philips', 'once')) || ~isempty(regexpi(xQ.Vendor, 'Siemens', 'once'))
+			xQ.PulseSequenceType = 'GRASE'; % assume that 3D Philips or Siemens is 3D GRASE
```
Again, I would make all lowercase, is more consistent and perhaps also BIDS compliant.

jan-petr replied:
3D and GRASE are BIDS compliant.

HenkMutsaerts commented on Functions/Administration/xASL_adm_DefineASLSequence.m:
```diff
@@ -99,17 +98,17 @@
 % readout-specific image processing for this sequence
-if ~isfield(xQ,'Sequence') || isempty(xQ.Sequence)
-    warning('No xQ.Sequence defined');
+if ~isfield(xQ, 'PulseSequenceType') || isempty(xQ.PulseSequenceType)
```
I should have changed this when making the population module sequence-independent.
Also note that this is not the same as checking Sequence (2d_epi vs epi).

We can/should remove all lines from 98 until 109.
If we cannot detect the PulseSequenceType and/or MRAcquisitionType, we simply give an error (we would never have this because we create ASL BIDS in our import)

1. In the ASL module, we always need x.Q.Readout. If it is missing, we can at the location where we need it (e.g., susceptibility mask creation) just not mask susceptibility -> the processing stays the same as what we do now, but without adding a potentially incorrect PulseSequenceType & MRAcquisitionType to the provenance.
2. In the population module, we assume that PulseSequenceType & MRAcquisitionType are in the json sidecar. If they are not (which should be a rare case), then we still not want incorrect PulseSequenceType & MRAcquisitionType but the effective resolution calculation part will give a warning and assume nominal resolution == effective resolution.

jan-petr replied:
Removed and we give an error.

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_ProcessM0.m:
```diff
@@ -146,7 +146,7 @@ function xASL_wrp_ProcessM0(x)
 if isfield(x.modules.asl, 'bRegisterM02ASL') && ~x.modules.asl.bRegisterM02ASL
     fprintf('M0 registration (to ASL or T1w) is skipped upon request\n');
-elseif ~strcmpi(x.Q.M0, 'UseControlAsM0') && isempty(regexpi(x.Q.Sequence, 'spiral'))
+elseif ~strcmpi(x.Q.M0, 'UseControlAsM0') && isempty(regexpi(x.Q.PulseSequenceType, 'spiral', 'once'))
```
So this has to be x.Q.Readout -> '3d_spiral' 

(note that we had this, assuming that x.Q.Sequence could only contain '2d_epi, 3d_grase, 3d_spiral, which we created ourselves. Now it can have 2d_spiral)

jan-petr replied:
Yes. I have changed this everywhere. But note that this old version was already looking for any 'spiral', e.g. 2D_spiral or 3D_spiral. So I can add 3D_spiral, but that would be changing the functionality compared to the older version.

HenkMutsaerts replied:
The point is that in the old version, we used a parameter created by ourselves (x.Q.Sequence), which could exclusively be 2d_epi, 3d_grase, 3d_spiral. Now, we have replaced this by a BIDS parameter that has multiple options. So we changed the functionality (in general, not in the most common cases).

So we can only replace checking x.Q.Sequence by checking both MRAcquisition & PulseSequenceType

jan-petr replied:
Yes, I understand. And I can add the 3D check here. But my point was that it's not really necessary here. As said - we check for dimension everywhere else. I thought adding here is not necessary, but I can do it if you want.

HenkMutsaerts replied:
So everywhere (not only here) where you renamed x.Q.Sequence to *PulseSequenceType, you potentially need to add the MRAcquisitionType check (even though you think it won't matter). That is my only wish ;)

HenkMutsaerts reviewed (changes requested):
Note that you cannot do a simple find/replace, because the value/meaning of PulseSequenceType (e.g. epi) is not identical to Sequence (e.g., 2d_epi).

We still need Sequence (e.g., 2d_epi), because many of our processing is dependent on it. You could simplify to epi grase spiral but you can have 2d spiral or 3d epi, even though it doesn't occur frequently, we need to catch this.

So I suggest the following:

1. readoutDim -> MRAcquisitionType so it is BIDS compliant (both json and internal)
2. x.Q.Sequence -> x.Q.Readout
3. we add PulseSequenceType

jan-petr commented:
> Note that you cannot do a simple find/replace, because the value/meaning of PulseSequenceType (e.g. epi) is not identical to Sequence (e.g., 2d_epi).
> 
> We still need Sequence (e.g., 2d_epi), because many of our processing is dependent on it. You could simplify to epi grase spiral but you can have 2d spiral or 3d epi, even though it doesn't occur frequently, we need to catch this.
> 
> So I suggest the following:
> 
> 1. readoutDim -> MRAcquisitionType so it is BIDS compliant (both json and internal)
> 2. x.Q.Sequence -> x.Q.Readout
> 3. we add PulseSequenceType

This is explained in the issue. MRAcquisitionType and PulseSequenceType are both BIDS fields. x.Q.Readout isn't. So no point introducing Readout when MRAcquisitionType and PulseSequenceType provide all the necessary information.

jan-petr committed:
#1674 Minor fixes

HenkMutsaerts reviewed (changes requested):
1. The point is that in the old version, we used a parameter created by ourselves (x.Q.Sequence), which could exclusively be 2d_epi, 3d_grase, 3d_spiral. Now, we have replaced this by a BIDS parameter that has multiple options. So we changed the functionality (in general, not in the most common cases).

So we can only replace checking x.Q.Sequence by checking both MRAcquisition & PulseSequenceType

So again, check throughout all cases where you changed x.Q.Sequence to PulseSequenceType, and see if this should not be MRAcqsuition & PulseSequenceType

2. This is all about the readout, not the sequence (sequence == labeling+readout). So calling it readout instead of sequence can be more clear.

I feel strong about point 1, not about point 2

jan-petr commented:
> 1. The point is that in the old version, we used a parameter created by ourselves (x.Q.Sequence), which could exclusively be 2d_epi, 3d_grase, 3d_spiral. Now, we have replaced this by a BIDS parameter that has multiple options. So we changed the functionality (in general, not in the most common cases).
> 
> So we can only replace checking x.Q.Sequence by checking both MRAcquisition & PulseSequenceType
> 
> So again, check throughout all cases where you changed x.Q.Sequence to PulseSequenceType, and see if this should not be MRAcqsuition & PulseSequenceType
> 
> 2. This is all about the readout, not the sequence (sequence == labeling+readout). So calling it readout instead of sequence can be more clear.
> 
> I feel strong about point 1, not about point 2

Point 1 -> this has been changed everywhere. We are checking both for readoutType and readoutDim everywhere. Except in one single case, where I think that checking for 3D is not necessary - see the comment above. But even there, I can add it if you want.

Point 2 -> It's a small change -> done

jan-petr committed:
#1674 Update Sequence to Readout
